lighthouse-core/scripts/lantern/minify-trace.js

```diff
@@ -0,0 +1,131 @@
+ */
+'use strict';
+
+/* eslint-disable no-console */
+
+const fs = require('fs');
+const path = require('path');
+
+if (process.argv.length !== 4) {
+  console.error('Usage $0: <input file> <output file>');
+  process.exit(1);
+}
+
+const inputTracePath = path.resolve(process.cwd(), process.argv[2]);
+const outputTracePath = path.resolve(process.cwd(), process.argv[3]);
+const inputTraceRaw = fs.readFileSync(inputTracePath, 'utf8');
+/** @type {LH.Trace} */
+const inputTrace = JSON.parse(inputTraceRaw);
+
+const toplevelTaskNames = new Set([
+  'TaskQueueManager::ProcessTaskFromWorkQueue',
+  'ThreadControllerImpl::DoWork',
+]);
+
+const traceEventsToAlwaysKeep = new Set([
+  'Screenshot',
+  'TracingStartedInBrowser',
+  'TracingStartedInPage',
+  'navigationStart',
+  'ParseAuthorStyleSheet',
+  'ParseHTML',
+  'PlatformResourceSendRequest',
+  'ResourceSendRequest',
+  'ResourceReceiveResponse',
+  'ResourceFinish',
+  'ResourceReceivedData',
+  'EventDispatch',
+]);
+
+const traceEventsToKeepInToplevelTask = new Set([
+  'TimerInstall',
+  'TimerFire',
+  'InvalidateLayout',
+  'ScheduleStyleRecalculation',
+  'EvaluateScript',
+  'XHRReadyStateChange',
+  'FunctionCall',
+  'v8.compile',
+]);
+
+const traceEventsToKeepInProcess = new Set([
+  ...toplevelTaskNames,
+  ...traceEventsToKeepInToplevelTask,
+  'firstPaint',
+  'firstContentfulPaint',
+  'firstMeaningfulPaint',
+  'firstMeaningfulPaintCandidate',
+  'loadEventEnd',
+  'domContentLoadedEventEnd',
+]);
+
+/**
+ * @param {LH.TraceEvent[]} events
+ */
+function filterTraceEvents(events) {
+  const startedInPageEvt = events.find(evt => evt.name === 'TracingStartedInPage');
+  if (!startedInPageEvt) throw new Error('Could not find TracingStartedInPage');
+
+  // Filter out event names we don't care about and tasks <1ms
+  let filtered = events.filter(evt => {
+    if (toplevelTaskNames.has(evt.name) && evt.dur < 1000) return false;
+    if (evt.pid === startedInPageEvt.pid && traceEventsToKeepInProcess.has(evt.name)) return true;
+    return traceEventsToAlwaysKeep.has(evt.name);
+  });
+
+  // Filter out events not inside a toplevel task
+  const toplevelRanges = filtered
+    .filter(evt => toplevelTaskNames.has(evt.name))
+    .map(evt => [evt.ts, evt.ts + evt.dur]);
+
+  /** @param {LH.TraceEvent} e */
+  const isInToplevelTask = e => toplevelRanges.some(([start, end]) => e.ts >= start && e.ts <= end);
+
+  filtered = filtered.filter((evt, index) => {
+    if (!traceEventsToKeepInToplevelTask.has(evt.name)) return true;
+    if (!isInToplevelTask(evt)) return false;
+
+    if (evt.ph === 'B') {
+      const endEvent = filtered.slice(index).find(e => e.name === evt.name && e.ph === 'E');
+      return endEvent && isInToplevelTask(endEvent);
+    } else {
+      return true;
+    }
+  });
+
+  // Filter down the screenshots to key moments + 2fps animations
+  const screenshotTimestamps = filtered.filter(evt => evt.name === 'Screenshot').map(evt => evt.ts);
```

Comment:
can you break each of these techniques into its own fn. i might want to use just the screenshots one in the future.

Reply:
👍 done